### PR TITLE
Data provider ratelimits

### PR DESCRIPTION
**Pull request by emmick4**
This PR adds the ability to reference declared data provider rate limits. Current notable missing functionality: manually overriding limits, defining custom enterprise limits.

### CODE REVIEW

File: ratelimits/package.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  "types": "dist/index.d.ts",
+  "files": [
+    "dist"
+  ],
+  "repository": {
+    "url": "https://github.com/smartcontractkit/external-adapters-js",
+    "type": "git"
+  },
+  "license": "MIT",
+  "scripts": {
```
Comment (justinkaseman) [inline ratelimits/package.json:21]: Won't need these:

`server`
`server:dist`
`start`

File: ratelimits/package.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@chainlink/ratelimits",
+  "version": "0.0.3",
```
Comment (justinkaseman) [inline ratelimits/package.json:3]:
```suggestion
  "version": "0.0.1",
```

File: ratelimits/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  const [burst, quota] = calculateLimits(declaredTier)
+  return {
+    burst: burst,
+    quota: quota,
+    second: burst,
+    minute: quota,
+  }
+}
+
+console.log(getRateLimit('blockchair', 5, undefined))
```
Comment (justinkaseman) [inline ratelimits/src/index.ts:71]:
```suggestion
```

File: ratelimits/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import Limits from './limits.json'
+const limits: Record<string, any> = Limits // ugly solution to dynamically reference JSON properties
```
Comment (justinkaseman) [inline ratelimits/src/index.ts:2]:
```suggestion
type provider = keyof typeof Limits;
```

File: ratelimits/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    tier !== undefined
+      ? limits[provider][tier]
+      : limits[provider].find((element: Record<string, unknown>) => element.tierName === tierName)
+  if (matchedTier === undefined) {
+    throw Error("tier or tierName doesn't match provider spec in limits.json")
+  }
+  return matchedTier
+}
+
+export const getRateLimit = (
```
Comment (justinkaseman) [inline ratelimits/src/index.ts:53]: Might be better DevEx to take a required `tier` and determine if it parses to an Int or is a string (name)

File: ratelimits/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  return [burst, quota]
+}
+
+const findTier = (provider: string, tier?: number, tierName?: string): DeclaredTier => {
+  const matchedTier =
+    tier !== undefined
+      ? limits[provider][tier]
+      : limits[provider].find((element: Record<string, unknown>) => element.tierName === tierName)
```
Comment (justinkaseman) [inline ratelimits/src/index.ts:46]:
```suggestion
      : (Limits[provider] as DeclaredTier[]).find( element => element.tierName === tierName )
```

File: ratelimits/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    },
+    {
+      "rateLimit1h":416666.66,
+      "tierName":"business+"
+    }
+  ], ...
+```
+- Each Object in the provider's array corresponds to a "tier" offered, in ascending order by price.
+- Valid rateLimit definitions are `rateLimit1s`, `rateLimit1m`, and `rateLimit1h`, corresponding to the closest time-based
+limit providers declare, 1 second, 1 minute, and 1 hour, respectively, currently using 10,000 /s to reflect unlimited.
```
Comment (justinkaseman) [inline ratelimits/README.md:32]: Might be more intuitive to have undefined key/val pairs or `Inf` as unlimited?

Comment (emmick4) [reply ratelimits/README.md:32]: 👍 opted for `-1`

File: ratelimits/src/index.ts (hunk 4 above)
Comment (emmick4) [reply ratelimits/src/index.ts]: couldn't sort the typing issue out here, wanted to be able to call `getRateLimit` with `tier` as a `string`

**Comment by justinkaseman** [comment]
LGTM. Just needs a rebase and then let's get this in to start using it with rate limiting.

**Pull request merged**